### ISSUE TITLE
Multiprocessing_train blocked

### ISSUE THREAD

**Issue by aslu98** [facebookresearch/ParlAI#2962]
I started a multiprocessing training with the following command:
`parlai multiprocessing_train -t fromfile:parlaiformat \
      --fromfile_datapath ../parlai_utilities/covid_qa+FAQ_parlaiformat.txt -m transformer/generator \
        --init-model zoo:blender/blender_3B/model --dict-file zoo:blender/blender_3B/model.dict \
        -veps 0.25 --attention-dropout 0.0 --batchsize 32 --embedding-size 2560 \
         --ffn-size 10240 --variant prelayernorm --n-heads 32 --n-positions 128 \
        --n-encoder-layers 2 --n-decoder-layers 24 \
        --history-add-global-end-token end --delimiter ' ' \
        --dict-tokenizer bytelevelbpe --dropout 0.1 --fp16 True \
        --label-truncate 128 --log_every_n_secs 10 -lr 7e-06 --lr-scheduler reduceonplateau \
        --lr-scheduler-patience 3 --optimizer adam --relu-dropout 0.0 --activation gelu \
        --model-parallel true --text-truncate 128 --truncate 128 \
        --warmup_updates 100 --fp16-impl mem_efficient --update-freq 2 --gradient-clip 0.1 \
        --skip-generation True -vp 10 -vmt ppl -vmm min \
        --save-after-valid True --model-file  covidmodel1`

Using two 12GB GPUs but it is has been stuck on `07:54:19 INFO | num words = 8008` for almost two hours.
I checked nvidia-smi and it is using the GPUs and `python -c 'import torch; print(torch.cuda.is_available())'` returns True.

multiprocessing_train output: https://imgur.com/T0GEUMl
nvidia-smi: https://imgur.com/CuREqRn

From the Recipe page I read that I could try to fine-tune 3B model with multiple GPUs, is it stuck because I don't have enough RAM? What else could it be? I already tried to stop the container and restart it.

**Comment by stephenroller** [comment]
You can’t mix multiprocessing train with model parallel true. The two processes just try to fight over both GPUs.

I’ll try to add an error message.

**Comment by aslu98** [comment]
Thanks! So for my understanding using the multiprocessing_train script is the same as using train_model with --model-parallel true, right?

**Comment by stephenroller** [comment]
They're not exactly the same. `mp_train` implements data parallelism, while `--model-parallel true` implements model parallelism. See here: https://leimao.github.io/blog/Data-Parallelism-vs-Model-Paralelism/. The tradeoff is roughly that data parallelism is faster but creates multiple copies of a model which uses more memory. Model parallelism is slower, but each GPU only uses the minimum amount of memory it needs, so you can train bigger models at the cost of speed.

While DP and MP are not mutually exclusive by design, the implementations of them in ParlAI are, because they both try to use 100% of the GPUs in your system. With them both on, these two parallelisms spend all their time fighting for the same resources.

**Comment by aslu98** [comment]
Thank you, now I understand what is going on much better.

**Comment by stephenroller** [comment]
FYI made a PR to make this give a loud error message.
